sdk/state/close.go

```diff
@@ -0,0 +1,120 @@
+	"github.com/stellar/experimental-payment-channels/sdk/txbuild"
+	"github.com/stellar/go/txnbuild"
+	"github.com/stellar/go/xdr"
+)
+
+// The steps for a channel coordinated close are as followed:
+// 1. Initiator or Responder submits latest declaration tx
+// (in steps 2-4 Initiator and Responder are interchangeable, as long as they alternate)
+// 2. Initiator calls ProposeCoordinatedClose
+// 3. Responder calls ConfirmCoordinatedClose
+// 4. Initiator calls ConfirmCoordinatedClose
+
+type CoordinatedClose struct {
+	observationPeriodTime      time.Duration
+	observationPeriodLedgerGap int64
+	closeSignatures            []xdr.DecoratedSignature
+}
+
+func (cc CoordinatedClose) CloseSignatures() []xdr.DecoratedSignature {
+	return cc.closeSignatures
+}
+
+func (c *Channel) CloseTxs() (txDecl *txnbuild.Transaction, txClose *txnbuild.Transaction, err error) {
+	txDecl, err = txbuild.Declaration(txbuild.DeclarationParams{
+		InitiatorEscrow:         c.initiatorEscrowAccount().Address,
+		StartSequence:           c.startingSequence,
+		IterationNumber:         c.latestCloseAgreement.IterationNumber,
+		IterationNumberExecuted: 0,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	txClose, err = c.makeCloseTx(c.observationPeriodTime, c.observationPeriodLedgerGap)
+	if err != nil {
+		return nil, nil, err
+	}
+	return txDecl, txClose, nil
+}
+
+func (c *Channel) CoordinatedCloseTx() (*txnbuild.Transaction, error) {
+	txClose, err := c.makeCloseTx(c.coordinatedClose.observationPeriodTime, c.coordinatedClose.observationPeriodLedgerGap)
+	if err != nil {
+		return nil, err
+	}
+	return txClose, nil
+}
+
+// ProposeCoordinatedClose proposes parameters for a close transaction to be submitted earlier.
+// This should be used when participants are in agreement on the final txClose parameters, but would
+// like to submit earlier than the original observation time.
+func (c *Channel) ProposeCoordinatedClose(observationPeriodTime time.Duration, observationPeriodLedgerGap int64) (CoordinatedClose, error) {
+	txCoordinatedClose, err := c.makeCloseTx(observationPeriodTime, observationPeriodLedgerGap)
+	if err != nil {
+		return CoordinatedClose{}, err
+	}
+	txCoordinatedClose, err = txCoordinatedClose.Sign(c.networkPassphrase, c.localSigner)
+	if err != nil {
+		return CoordinatedClose{}, nil
+	}
+	return CoordinatedClose{
+		observationPeriodTime:      observationPeriodTime,
+		observationPeriodLedgerGap: observationPeriodLedgerGap,
+		closeSignatures:            txCoordinatedClose.Signatures(),
+	}, nil
+}
+
+func (c *Channel) ConfirmCoordinatedClose(cc CoordinatedClose) (CoordinatedClose, error) {
+	txCoordinatedClose, err := c.makeCloseTx(cc.observationPeriodTime, cc.observationPeriodLedgerGap)
+	if err != nil {
+		return CoordinatedClose{}, err
+	}
+
+	// If remote has not signed coordinated close, error as is invalid.
+	signed, err := c.verifySigned(txCoordinatedClose, cc.closeSignatures, c.remoteSigner)
+	if err != nil {
+		return CoordinatedClose{}, err
```

Comment:
We should wrap `err` here with `fmt.Errorf` like we do in other places so that if an error occurs we know what was happening when it occurred.

E.g.: https://github.com/stellar/experimental-payment-channels/blob/c30a604c216a5d8015e20309f3aad4d20ef67fb6/sdk/state/open.go#L120

Reply:
https://github.com/stellar/experimental-payment-channels/pull/95/commits/ac3ee05b8c4e42b0d3eec6522f5d4339dd26c6d2